src/_pytest/python_api.py

```diff
@@ -668,29 +707,45 @@ def raises(expected_exception, *args, match=None, **kwargs):
         try:
             func(*args[1:], **kwargs)
         except expected_exception:
-            return _pytest._code.ExceptionInfo.from_current()
+            # Cast to narrow the type to expected_exception (_E).
+            return cast(
+                _pytest._code.ExceptionInfo[_E],
+                _pytest._code.ExceptionInfo.from_current(),
+            )
     fail(message)
 
 
 raises.Exception = fail.Exception  # type: ignore
 
 
-class RaisesContext:
-    def __init__(self, expected_exception, message, match_expr):
+class RaisesContext(Generic[_E]):
+    def __init__(
+        self,
+        expected_exception: Union["Type[_E]", Tuple["Type[_E]", ...]],
+        message: str,
+        match_expr: Optional[Union[str, Pattern]] = None,
+    ) -> None:
         self.expected_exception = expected_exception
         self.message = message
         self.match_expr = match_expr
-        self.excinfo = None
+        self.excinfo = None  # type: Optional[_pytest._code.ExceptionInfo[_E]]
 
-    def __enter__(self):
+    def __enter__(self) -> _pytest._code.ExceptionInfo[_E]:
         self.excinfo = _pytest._code.ExceptionInfo.for_later()
         return self.excinfo
 
-    def __exit__(self, *tp):
+    def __exit__(
+        self,
+        exc_type: Optional["Type[BaseException]"],
+        exc_val: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
+    ) -> bool:
         __tracebackhide__ = True
-        if tp[0] is None:
+        if exc_type is None:
             fail(self.message)
-        self.excinfo.__init__(tp)
+        assert self.excinfo is not None
+        # Type ignored because mypy doesn't like calling __init__ directly like this.
+        self.excinfo.__init__((exc_type, exc_val, exc_tb))  # type: ignore
```

Comment:
should we make a separate initialization function for this? I'd rather not `type: ignore` where we control both sides of the code

Reply:
That would certainly be less weird, I'll do that.